AkitaInteractive/Hachiko-Engine
# Editor Support for Scripts

This pull request adds pre-build time automatically generated editor content support for scripts. However there are things to be careful about.

## Supported Types:
- int, 
- unsigned(use unsigned instead of "unsigned int" as current parser does not support two worded types), 
- float, 
- double, 
- bool, 
- std::string, 
- GameObject*, 
- any Component* (including scripts), 
- math::float2, 
- math::float3,
- math::float4.

Although the exposing any ```Script*``` to editor if declared with ```SERIALIZE_FIELD(, )``` macro is supported, with the current architecture we don't support runtime serialization of ```Script*``` due to memory location of scripts belonging to the dll. But this will be fixed for convenience of the gameplay programmers in the following pr's as I have the solution in mind. Any other type than listed above is currently not supported, but, extending this list is not an issue at all, so in case of any need let's discuss and add those as well. Perhaps a cool addition can be to add std::vector, given that the vector stores the types I've listed above.

## How to Test?
1. Clean & Build Hachiko
2. Delete all the lines of Gameplay/generated/GeneratedEditor.cpp, Gameplay/generated/SerializationMethods.cpp, and Gameplay/generated/Factory.cpp
3. Clean & Build Gameplay.
4. See if there are changes to the files described at 2.
5. Run the engine.
6. Add a script to a GameObject.
7. Add/remove fields of that Script which are surrounded by ```SERIALIZE_FIELD``` macro.
9. Hit the play button on the editor.
10. While the game is running interact with fields.
11. Try dragging and dropping GameObjects to the ```SERIALIZED_FIELD```s with types of either ```GameObject*``` or any ```Component*```
12. If you feel fancy, write some scripts that have ```nullptr``` serialized fields of ```GameObject*``` or ```Component*```, have them ```if```s for ```nullptr``` check of those fields and if not ```nullptr``` logging the name of the component/gameobject or some fields in the ```OnUpdate``` method.

## Things That Will Be Added in the Upcoming PRs
- Generated ```Save``` and ```Load``` methods for scripts. This change will make scripting come close to the real Hachiko engine experience as right now we lack saving the values of fields that are surrounded by ```SERIALIZE_FIELD```.
- Script* serialization. (Probably far more later than anything else since we can just add ```GameObject*``` to the script and do get component on the ```OnAwake``` method for the script we want to get).
